ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/trielog/TrieLogPruner.java

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright contributors to Hyperledger Besu.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.hyperledger.besu.ethereum.bonsai.trielog;
+
+import org.hyperledger.besu.datatypes.Hash;
+import org.hyperledger.besu.ethereum.bonsai.storage.BonsaiWorldStateKeyValueStorage;
+import org.hyperledger.besu.ethereum.chain.Blockchain;
+import org.hyperledger.besu.ethereum.core.BlockHeader;
+import org.hyperledger.besu.ethereum.core.ProcessableBlockHeader;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Stream;
+
+import com.google.common.collect.Multimap;
+import com.google.common.collect.TreeMultimap;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TrieLogPruner {
+
+  private static final Logger LOG = LoggerFactory.getLogger(TrieLogPruner.class);
+
+  public static final long DEFAULT_RETENTION_THRESHOLD = 0L;
+  public static int DEFAULT_PRUNING_LIMIT = 30_000;
+
+  private final int pruningLimit;
+  private final int loadingLimit;
+  private final BonsaiWorldStateKeyValueStorage rootWorldStateStorage;
+  private final Blockchain blockchain;
+  private final long numBlocksToRetain;
+  private static final Multimap<Long, byte[]> knownTrieLogKeysByDescendingBlockNumber =
+      TreeMultimap.create(Comparator.reverseOrder(), Comparator.comparingInt(Arrays::hashCode));
+
+  public TrieLogPruner(
+      final BonsaiWorldStateKeyValueStorage rootWorldStateStorage,
+      final Blockchain blockchain,
+      final long numBlocksToRetain,
+      final int pruningLimit) {
+    this.rootWorldStateStorage = rootWorldStateStorage;
+    this.blockchain = blockchain;
+    this.numBlocksToRetain = numBlocksToRetain;
+    this.pruningLimit = pruningLimit;
+    this.loadingLimit = pruningLimit; // same as pruningLimit for now
+  }
+
+  public void initialize() {
+    preloadCache();
+  }
+
+  private void preloadCache() {
+    LOG.atInfo()
```

Comment:
do we need info for that ? debug is not enough ?

Reply:
This is because it introduces a small delay (currently tested as 4-8 seconds with default limit of 30,000) 
so I wanted to indicate to the user what the delay was for...similar to the "compacting database..." message